casper/src/main/scala/coop/rchain/casper/engine/BlockReceiverState.scala

```diff
@@ -0,0 +1,18 @@
+package coop.rchain.casper.engine
+
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+
+sealed trait RecvStatus
+case object RecvStarted  extends RecvStatus
+case object RecvFinished extends RecvStatus
+
+final case class BlockReceiverState[F[_], MId](state: Map[MId, RecvStatus]) {
+  def contains(m: MId): Boolean       = state.contains(m)
+  def set(m: MId, status: RecvStatus) = new BlockReceiverState[F, MId](state + (m -> status))
+}
+
+object BlockReceiverState {
+  def create[F[_]: Sync, MId]: F[Ref[F, BlockReceiverState[F, MId]]] =
+    Ref.of(BlockReceiverState[F, MId](Map.empty[MId, RecvStatus]))
+}
```

Comment:
Please take a look how LfsBlockRequester is defined with its internal state inside Ref which is created in a function which wraps the whole component and returns Stream of received block hashes.

https://github.com/rchain/rchain/blob/dd46a6b59f0d7585ae27148f72b2c236197b0bb3/casper/src/main/scala/coop/rchain/casper/engine/LfsBlockRequester.scala#L139-L351